stock_intercompany/models/stock_picking.py

```diff
@@ -1,49 +1,270 @@
+            self.env["res.company"]
+            .sudo()
+            .search([("partner_id", "=", self.partner_id.id)], limit=1)
+        )
+        # Only create a picking if the recipient is a different company
+        if not company or company == self.company_id:
+            return
+
+        # Restrict to the configured mode on the target company
+        if company.intercompany_picking_creation_mode not in [mode, "both"]:
+            return
+
+        if not self._check_intercompany_company(company, mode):
+            return
+
+        return company
+
+    def _create_counterpart_picking(self, mode):
+        """
+        Create a counterpart picking for the given picking in the given mode:
+        Mode in: Create incoming from outgoing picking
+        Mode out: Create outgoing from incoming picking
+        """
+
+        self.ensure_one()
+        if self.state == "cancel":
+            return
+
+        company = self._can_create_counterpart(mode)
+        if not company:
+            return
+
+        intercompany_type = getattr(company, "intercompany_%s_type_id" % mode)
+
+        warehouse = intercompany_type.warehouse_id or (
+            self.env["stock.warehouse"]
+            .sudo()
+            .search([("company_id", "=", company.id)], limit=1)
+        )
+
+        # Defalut intercompany type to warehouse in/out type
+        intercompany_type = intercompany_type or getattr(warehouse, "%s_type_id" % mode)
+
+        new_picking_vals = self._prepare_counterpart_picking_vals(
+            company, intercompany_type, warehouse, mode
+        )
+        picking = self.env["stock.picking"].sudo().create(new_picking_vals)
+        self.is_locked = True
+        picking.action_confirm()
+        return picking
+
+    def _check_intercompany_company(self, company, mode):
+        """
+        Hook to check if the given company is valid for the given mode.
+        If you override this method, please update the domain
+        """
+        return True
+
+    def _prepare_counterpart_picking_vals(
+        self, company, intercompany_type, warehouse, mode
+    ):
+        """Prepare the values to create the counterpart picking"""
+
+        # in : suppliers -> stock
+        # out: stock -> customers
+        location_src = (
+            self.env.ref("stock.stock_location_suppliers")
+            if mode == "in"
+            else warehouse.lot_stock_id
+        )
+        location_dest = (
+            warehouse.lot_stock_id
+            if mode == "in"
+            else self.env.ref("stock.stock_location_customers")
+        )
+
+        vals = {
+            "partner_id": self.company_id.partner_id.id,
+            "company_id": company.id,
+            "picking_type_id": intercompany_type.id,
+            "state": "draft",
+            "location_id": location_src.id,
+            "location_dest_id": location_dest.id,
+            "intercompany_parent_id": self.id,  # Keep track of the parent picking
+        }
```

Comment:
Please set group_id: False in the counterpart picking. procurement groups are not shared between companies.